### PR TITLE
ARROW-7329: [Java] AllocationManager: Allow managing different types …

### PR DESCRIPTION
**Pull request by zhztheplayer**
…of memory other than those are allocated using Netty

See https://jira.apache.org/jira/browse/ARROW-7329

### CODE REVIEW

**Comment by jacques-n** [comment]
Nice work. Can you make a change so that allocationmanager is the base class and NettyAllocationManager is the subclass? Will make this much easier to review and track history.

**Comment by zhztheplayer** [comment]
@jacques-n - Sure, will do. I was to avoid breaking change to AllocationManager, but I see we have  commented:
https://github.com/apache/arrow/blob/6cfe3e08a4c0100716e3f34325a113bffb2f47a1/java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java#L36-L38
So actually I am not sure if it's necessary to preserve the class name. Thanks for your suggestion.

File: java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java [hunk 1]
```diff
@@ -523,7 +522,7 @@ void print(StringBuilder sb, int indent, BaseAllocator.Verbosity verbosity) {
     }
   }
 
-  public UnsafeDirectLittleEndian getUnderlying() {
-    return allocationManager.getMemoryChunk();
+  public AllocationManager getAllocationManager() {
```
Comment (zhztheplayer) [inline java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: This is breaking but the method seems to be used only in test. Any suggestions?

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java [hunk 2]
```diff
@@ -172,10 +154,10 @@ void release(final BufferLedger ledger) {
         // the only <allocator, reference manager> mapping was for the owner
         // which now has been removed, it implies we can safely destroy the
         // underlying memory chunk as it is no longer being referenced
-        ((BaseAllocator)oldLedger.getAllocator()).releaseBytes(size);
+        ((BaseAllocator)oldLedger.getAllocator()).releaseBytes(getSize());
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: why the change?

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java [hunk 3]
```diff
@@ -198,9 +180,20 @@ void release(final BufferLedger ledger) {
...
+
+  /**
+   * Return the absolute memory address pointing to the fist byte of underling memory chunk.
+   */
+  abstract long memoryAddress();
+
+  /**
+   * Release the underling memory chunk.
+   */
+  abstract void release0();
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: protected?

File: java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java [hunk 4]
```diff
@@ -355,6 +354,17 @@ private ArrowBuf bufferWithoutReservation(
     return buffer;
   }
 
+  private AllocationManager newAllocationManager(int size) {
+    return newAllocationManager(this, size);
+  }
+
+  /**
+   * All {@link AllocationManager} instances created are defaulted to Netty implementation.
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java]: comment arguemnts

File: java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java [hunk 5]
```diff
@@ -141,7 +140,7 @@ public boolean release(int decrement) {
    * no ArrowBufs managed by this reference manager need access to the memory
    * chunk. In that case, the ledger should inform the allocation manager
    * about releasing its ownership for the chunk. Whether or not the memory
-   * chunk will be released is something that {@link AllocationManager} will
+   * chunk will be released is something that {@link NettyAllocationManager} will
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: why change?

File: java/memory/src/main/java/org/apache/arrow/memory/ChildAllocator.java [hunk 6]
```diff
@@ -53,5 +53,9 @@
     super(parentAllocator, listener, name, initReservation, maxAllocation, roundingPolicy);
   }
 
-
+  @Override
+  protected AllocationManager newAllocationManager(BaseAllocator accountingAllocator, int size) {
+    // to expect consistent behavior with parent on creating allocation managers
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/ChildAllocator.java]: This is kind of wonky. Let's have a factory instead.

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java [hunk 7]
```diff
@@ -47,22 +43,17 @@
  * typical query. The
  * contention of acquiring a lock on AllocationManager should be very low.
  */
-public class AllocationManager {
+public abstract class AllocationManager {
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: rather than making this abstract, let's change to a BufferAllocator configuration that includes a AllocationManager factory. That way we avoid breaking encapsulation and can also have other configuration options carried across allocators.

File: java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java (hunk 5 above)
Comment (zhztheplayer) [reply java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: Seems to be mistake. Thanks for catch.

File: java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java (hunk 4 above)
Comment (zhztheplayer) [reply java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java]: Done.

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java (hunk 2 above)
Comment (zhztheplayer) [reply java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: This is to keep up with the current behaviour. Following is the constructor code of current `AllocationManager`:
https://github.com/apache/arrow/blob/21a4ffe7411ac62c97d636a478481d96c26380a3/java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java#L73-L84

This allows an AllocationManager allocates different bytes of memory comparing to desired. So at here, `size` is desired size passed by constructor, while `getSize()` should be the actual size. I'll do some name changes in case of confusing.

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java (hunk 3 above)
Comment (zhztheplayer) [reply java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: I was thinking to force the implementations to be packaged under `org.apache.arrow.memory`... Which doesn't seem to be necessary.

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java (hunk 7 above)
Comment (zhztheplayer) [reply java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: Thanks! It's reasonable but AllocationManager's constructor requires an `BaseAllocator` instance as a parameter, as a result, passing a AllocationManager factory to BufferAllocator's constructor brings dependency cycle (chicken & egg). Should we try decouple `BaseAllocator` from AllocationManager's constructor?

And probably I am not able to correctly getting this:
> rather than making this abstract, ...

Does this mean we should keep the `public class AllocationManager`, and make it extend a base class? Like what's in my first version?

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java (hunk 3 above)
Comment (zhztheplayer) [reply java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: Done. Also changed `memoryAddress()` and constructor.

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java (hunk 7 above)
Comment (jacques-n) [reply java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: Why not just have the factory accept all the required items? Is there a reason this won't work?

Comment (zhztheplayer) [reply java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: It seems a lambda-like factory should work properly, though I forgot what stopped me when I try using a factory at first. I'll duble-check later :)

Comment (zhztheplayer) [reply java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: @jacques-n - Just changed to a config/factory approach. Could you please take a look? Thanks a lot.

File: java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java (hunk 1 above)
Comment (jacques-n) [reply java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: Let's add a new method which is 
T unwrap(Class<T> clazz) 

Then mark getUnderlying as deprecated and have it delegatee to unwrap(UnsafeDirectLittleEndian.class) and then also support unwrap(AllocationManager.class).

File: java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java [hunk 8]
```diff
@@ -729,6 +745,146 @@ private void dumpBuffers(final StringBuilder sb, final Set<BufferLedger> ledgerS
...
+   */
+  public static Config defaultConfig() {
+    return DEFAULT_CONFIG;
+
+  }
+
+  /**
+   * Returns a builder class for configuring BaseAllocator's options.
+   */
+  public static ConfigBuilder configBuilder() {
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java]: Can you add [Immutables]( http://immutables.github.io/) to the project and use that instead. Should allow us to avoid creation/management of builders. I've started using it in another project and found it quite helpful at avoiding lots of scaffolding code.

File: java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java [hunk 9]
```diff
@@ -190,7 +190,7 @@ public void testAllocator_getParentAndChild() throws Exception {
 
   @Test
   public void testAllocator_childRemovedOnClose() throws Exception {
-    try (final RootAllocator rootAllocator = new RootAllocator(MAX_ALLOCATION)) {
+    try (final RootAllocator rootAllocator = createRootAllocator(MAX_ALLOCATION)) {
```
Comment (jacques-n) [inline java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java]: let's avoid this change

Comment (zhztheplayer) [reply java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java]: Agreed. Something is over-tested.

File: java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java (hunk 8 above)
Comment (zhztheplayer) [reply java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java]: > Let's add a new method which is
> T unwrap(Class clazz)

Done.

> Can you add Immutables to the project and use that instead.

Thanks for suggesting Immutables, I have tried and it works pretty well. The only thing I am worried about is it slightly increases the development complexity. But I know a lot of projects already deal with the conditions in similar way so I have no objections on introducing it. 
Done.

File: java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java (hunk 9 above)
Comment (zhztheplayer) [reply java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java]: Done.

**Review by jacques-n — APPROVED** [review]
Sorry I missed coming back to this. This LGTM. Thanks for pulling it to completion.

+1

File: java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java [hunk 10]
```diff
@@ -111,15 +112,15 @@ protected BaseAllocator(
...
-    this.roundingPolicy = roundingPolicy;
+    this.roundingPolicy = config.getRoundingPolicy();
   }
 
   AllocationListener getListener() {
     return listener;
   }
 
   @Override
-  public BufferAllocator getParentAllocator() {
```
Comment (lidavidm) [inline java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java]: Just to double-check - is it intentional that we're exposing the concrete implementation now here?

File: java/memory/pom.xml [hunk 11]
```diff
@@ -37,6 +37,10 @@
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-api</artifactId>
     </dependency>
+    <dependency>
```
Comment (lidavidm) [inline java/memory/pom.xml]: Is it worth taking a dependency for a single builder? At the very least, let's file a follow-up to see if we can replace any other builders in the project with Immutables as well.

**Review by lidavidm — COMMENTED** [review]
Just some quick general questions - I'm willing to go ahead and merge this though if nobody else gets to it.

File: java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java [hunk 12]
```diff
@@ -17,14 +17,12 @@
 
 package org.apache.arrow.memory;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNotEquals;
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
+
+import static org.junit.Assert.*;
```
Comment (liyafan82) [inline java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java]: Please expand this to avoid style errors.

File: java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java (hunk 10 above)
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java]: IMO, it is useful to expose the hierarchical structure of the allocators.

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java (hunk 2 above)
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: If size and getSize() are different, maybe this should be made explicit in the JavaDoc?

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+ * The default implementation of AllocationManagerBase. The implementation is responsible for managing when memory
+ * is allocated and returned to the Netty-based PooledByteBufAllocatorL.
+ */
+public class NettyAllocationManager extends AllocationManager {
+
+  public static final Factory FACTORY = new Factory();
+
+  private static final PooledByteBufAllocatorL INNER_ALLOCATOR = new PooledByteBufAllocatorL();
+  static final UnsafeDirectLittleEndian EMPTY = INNER_ALLOCATOR.empty;
+  static final long CHUNK_SIZE = INNER_ALLOCATOR.getChunkSize();
```
Comment (liyafan82) [inline java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java]: It seems this is not used.

File: java/memory/pom.xml (hunk 11 above)
Comment (zhztheplayer) [reply java/memory/pom.xml]: I second the idea to factor other builders. So far I didn't have a check but we may make relevant changes in other issue/pr.

And I don't think the new dependency matters much as it's a single jar library and it is not needed in runtime.

File: java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java (hunk 12 above)
Comment (zhztheplayer) [reply java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java]: OK but It's weird that I didn't see any style errors on this.

Comment (zhztheplayer) [reply java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java]: Done

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java (hunk 13 above)
Comment (zhztheplayer) [reply java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java]: Actually it is used in DefaultRoundingPolicy via reflection:
https://github.com/apache/arrow/blob/7bd4e73513e1eef5287452dd18cbd808027f09db/java/memory/src/main/java/org/apache/arrow/memory/rounding/DefaultRoundingPolicy.java#L39-L47

**Review by liyafan82 — APPROVED** [review]

File: java/memory/pom.xml (hunk 11 above)
Comment (jacques-n) [reply java/memory/pom.xml]: I think this would be useful elsewhere. For example, we should really use it for the *.pojo classes. Also, as mentioned, this isn't used/required at runtime

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java (hunk 2 above)
Comment (jacques-n) [reply java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: yeah, maybe call one requested_size and one size (or similar)

File: java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java [hunk 14]
```diff
@@ -111,15 +112,15 @@ protected BaseAllocator(
...
+  public BaseAllocator getParentAllocator() {
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java]: It is okay to expose here as long as the BufferAllocator interface is BufferAllocator.

**Review by jacques-n — COMMENTED** [review]
I think this looks good. The size thing could be dealt with in a follow-on.

+1

File: java/memory/pom.xml (hunk 11 above)
Comment (lidavidm) [reply java/memory/pom.xml]: Ah - my mistake, non-runtime dependencies are even better.

**Comment by lidavidm** [comment]
I'll go ahead and merge, then file an issue to clarify the naming. Thanks @zhztheplayer for carrying this through!

**Comment by lidavidm** [comment]
I filed https://issues.apache.org/jira/browse/ARROW-7538 for the actual vs desired size clarification.

**Comment by zhztheplayer** [comment]
Thanks for everyone's review!
